crates/rpc/rpc/src/eth/gas_oracle.rs

```diff
@@ -0,0 +1,255 @@
+    pub max_price: Option<U256>,
+
+    /// The minimum gas price, under which the sample will be ignored
+    pub ignore_price: Option<U256>,
+}
+
+impl Default for GasPriceOracleConfig {
+    fn default() -> Self {
+        GasPriceOracleConfig {
+            blocks: 20,
+            percentile: 60,
+            max_header_history: 1024,
+            max_block_history: 1024,
+            default: None,
+            max_price: Some(DEFAULT_MAX_PRICE),
+            ignore_price: Some(DEFAULT_IGNORE_PRICE),
+        }
+    }
+}
+
+/// Calculates a gas price depending on recent blocks.
+#[derive(Debug)]
+pub struct GasPriceOracle<Client> {
+    /// The type used to subscribe to block events and get block info
+    client: Client,
+    /// The cache for blocks
+    cache: EthStateCache,
+    /// The config for the oracle
+    oracle_config: GasPriceOracleConfig,
+    /// The latest calculated price and its block hash
+    last_price: Mutex<GasPriceOracleResult>,
+}
+
+impl<Client> GasPriceOracle<Client>
+where
+    Client: BlockProviderIdExt + 'static,
+{
+    /// Creates and returns the [GasPriceOracle].
+    pub fn new(
+        client: Client,
+        mut oracle_config: GasPriceOracleConfig,
+        cache: EthStateCache,
+    ) -> Self {
+        // sanitize the perentile to be less than 100
+        if oracle_config.percentile > 100 {
+            warn!(prev_percentile=?oracle_config.percentile, "Invalid configured gas price percentile, using 100 instead");
+            oracle_config.percentile = 100;
+        }
+
+        Self { client, oracle_config, last_price: Default::default(), cache }
+    }
+
+    /// Suggests a gas price estimate based on recent blocks, using the configured percentile.
+    pub async fn suggest_tip_cap(&self) -> EthResult<U256> {
+        let header_hash = self
+            .client
+            .block_hash_for_id(BlockId::Number(BlockNumberOrTag::Latest))?
+            .ok_or(EthApiError::UnknownBlockNumber)?;
+
+        let mut last_price = self.last_price.lock().await;
+
+        // if we have stored a last price, then we check whether or not it was for the same head
+        if last_price.block_hash == header_hash {
+            return Ok(last_price.price)
+        }
+
+        // if all responses are empty, then we can return a maximum of 2*check_block blocks' worth
+        // of prices
+        //
+        // we only return more than check_block blocks' worth of prices if one or more return empty
+        // transactions
+        let mut current_hash = header_hash;
+        let mut results = Vec::new();
+        let mut populated_blocks = 0;
+
+        // we only check a maximum of 2 * max_block_history
+        for _ in 0..self.oracle_config.max_block_history * 2 {
+            let (parent_hash, block_values) = self
+                .get_block_values(current_hash, SAMPLE_NUMBER as usize)
+                .await?
+                .ok_or(EthApiError::UnknownBlockNumber)?;
```

Comment:
this can happen if the chain is not long enough yet, so `max_block_history` longer than chain length.

so we should probably cap this with header number

`self.oracle_config.max_block_history * 2`

however, this isn't really an issue because this edge case naturally goes away

Reply:
makes sense, bounded by block num